api/grpcserver/mesh_service.go

```diff
@@ -0,0 +1,495 @@
+	log.Info("GRPC MeshService.AccountMeshDataQuery")
+
+	startLayer := types.LayerID(in.MinLayer)
+
+	if startLayer > s.Mesh.LatestLayer() {
+		return nil, status.Errorf(codes.InvalidArgument, "`LatestLayer` must be less than or equal to latest layer")
+	}
+	if in.Filter == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter` must be provided")
+	}
+	if in.Filter.AccountId == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter.AccountId` must be provided")
+	}
+
+	// Gather transaction data
+	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	res := &pb.AccountMeshDataQueryResponse{}
+	if in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_TRANSACTIONS) != 0 {
+		meshTxIds := s.getTxIdsFromMesh(startLayer, addr)
+		mempoolTxIds := s.Mempool.GetTxIdsByAddress(addr)
+
+		// Look up full data for all unique txids
+		txs, missing := s.Mesh.GetTransactions(append(meshTxIds, mempoolTxIds...))
+
+		// TODO: Do we ever expect txs to be missing here?
+		// E.g., if this node has not synced/received them yet.
+		if len(missing) != 0 {
+			log.Error("could not find transactions %v", missing)
+			return nil, status.Errorf(codes.Internal, "error retrieving tx data")
+		}
+		for _, t := range txs {
+			res.Data = append(res.Data, &pb.AccountMeshData{
+				DataItem: &pb.AccountMeshData_Transaction{
+					Transaction: convertTransaction(t),
+				},
+			})
+		}
+	}
+
+	// Gather activation data
+	if in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_ACTIVATIONS) != 0 {
+		// We have no way to look up activations by coinbase so we have no choice
+		// but to read all of them.
+		// TODO: index activations by layer (and maybe by coinbase)
+		var activations []types.ATXID
+		for l := startLayer; l <= s.Mesh.LatestLayer(); l++ {
+			layer, err := s.Mesh.GetLayer(l)
+			if layer == nil || err != nil {
+				return nil, status.Errorf(codes.Internal, "error retrieving layer data")
+			}
+
+			for _, b := range layer.Blocks() {
+				activations = append(activations, b.ATXIDs...)
+			}
+		}
+
+		// Look up full data
+		atxs, matxs := s.Mesh.GetATXs(activations)
+		if len(matxs) != 0 {
+			log.Error("could not find activations %v", matxs)
+			return nil, status.Errorf(codes.Internal, "error retrieving activations data")
+		}
+		for _, atx := range atxs {
+			// Filter here, now that we have full data
+			if atx.Coinbase == addr {
+				pbatx, err := convertActivation(atx)
+				if err != nil {
+					return nil, status.Errorf(codes.Internal, "error serializing activation data")
+				}
+				res.Data = append(res.Data, &pb.AccountMeshData{
+					DataItem: &pb.AccountMeshData_Activation{
+						Activation: pbatx,
+					},
+				})
+			}
+		}
+	}
+
+	res.TotalResults = uint32(len(res.Data))
+
+	// Skip to offset, don't send more than max results
```

Comment:
I want to highlight the TODO item following this - another good example for a need for more optimization. CC @antonlerner